### ISSUE TITLE
syncoid with recursive raw send do not maintain encryptionroot

### ISSUE THREAD

**Issue by dani** [jimsalterjrs/sanoid#614]
Hi.
I'm using syncoid to replicate a encrypted dataset, with several (also encrypted) child datasets. On the source, the encryption root is zpool/backups. With childrens being zpool/backups/backuppc, zpool/backups/pbs etc. All with encryptionroot=zpool/backups

I'm replicating it on a remote target with :
```
/sbin/syncoid --identifier=backups --force-delete --compress=none --recursive  --sshcipher=aes128-ctr --sendoptions="w" zpool/backups zfs-recv@10.29.3.27:zpool/backups
```

Everything is correctly replicated, except that all the children are sent separately, so they all become their own encryptionroot on the dest, eg, on source:
```
[root@bkp3 ~]# zfs get encryptionroot zpool/backups/backuppc
NAME                    PROPERTY        VALUE          SOURCE
zpool/backups/backuppc  encryptionroot  zpool/backups  -
[root@bkp3 ~]#
```
On dest :
```
[root@bkp4 ~]$ zfs get encryptionroot zpool/backups/backuppc
NAME                    PROPERTY        VALUE                   SOURCE
zpool/backups/backuppc  encryptionroot  zpool/backups/backuppc  -
[root@bkp4 ~]$ 
```
Is there a way with syncoid to replicate a raw, encrypted sub-tree while maintaining the encryptionroot ?

**Comment by phreaker0** [comment]
@dani you need to let zfs take care of the recursion instead of syncoid, in this case you can't use syncoid stuff like "--exclude" and so so but in your case this doesn't matter, from the manpage of zfs:

```
       -R, --replicate
           Generate a replication stream package, which will replicate the specified file system, and all descendent file systems, up to the named snapshot.  When received, all properties, snap‐
           shots, descendent file systems, and clones are preserved.

           If the -i or -I flags are used in conjunction with the -R flag, an incremental replication stream is generated.  The current values of properties, and current snapshot and file system
           names are set when the stream is received.  If the -F flag is specified when this stream is received, snapshots and file systems that do not exist on the sending side are destroyed.
           If the -R flag is used to send encrypted datasets, then -w must also be specified.

```

so in your case this should work:

    /sbin/syncoid --identifier=backups --force-delete --compress=none --sshcipher=aes128-ctr --sendoptions="Rw" zpool/backups zfs-recv@10.29.3.27:zpool/backups

**Comment by dani** [comment]
Ok, makes sens. Thanks 🙂

**Comment by dani** [comment]
Unfortunatly, it's not working as expected. encryptionroot is still "reset" for each sub dataset. In anycase, as sanoid and syncoid don't take recursive snapshots, but only per data-set, I guess incremental send wouldn't work anyway

**Comment by phreaker0** [comment]
@dani you need to redo the initial replication to preserve the encryption root with the -R option.
Incremental send should work, recursive snapshots are not strictly needed in the sense of being atomic.
(Sanoid can take atomic recursive snapshots with 'recursive = zfs')

**Comment by dani** [comment]
I've tried to redo the initial repl, but it still created zpool/backups/backuppc with its own encryption root on the dest. I'm playing with smaller datasets now to understand what's needed to get it working (as my real backup dataset is ~18TB, it's not easy to play with)

**Comment by phreaker0** [comment]
@dani I just tried it and it works as expected for me, you just need to take an recursive snapshot for the initial replication, after that you can use the orindary syncoid recursion:

```
+ POOL_IMAGE=/tmp/syncoid-test-0-0.zpool
+ POOL_SIZE=1024M
+ POOL_NAME=syncoid-test-0-0
+ POOL_IMAGE2=/tmp/syncoid-test-0-1.zpool
+ POOL_SIZE2=1024M
+ POOL_NAME2=syncoid-test-0-1
+ truncate -s 1024M /tmp/syncoid-test-0-0.zpool
+ truncate -s 1024M /tmp/syncoid-test-0-1.zpool
+ zpool create -m none -f -O encryption=aes-256-gcm -O keylocation=prompt -O keyformat=passphrase syncoid-test-0-0 /tmp/syncoid-test-0-0.zpool
Enter passphrase:
Re-enter passphrase:
+ zpool create -m none -f syncoid-test-0-1 /tmp/syncoid-test-0-1.zpool
+ trap cleanUp EXIT
+ zfs create syncoid-test-0-0/a
+ zfs create syncoid-test-0-0/a/aa
+ zfs create syncoid-test-0-0/a/ab
+ zfs create syncoid-test-0-0/a/ac
+ zfs create syncoid-test-0-0/b
+ zfs create syncoid-test-0-0/c
+ zfs snapshot -r syncoid-test-0-0@replication
+ ../../../syncoid --no-sync-snap --identifier=backups --force-delete --compress=none --sendoptions=Rw syncoid-test-0-0 syncoid-test-0-1/backup
NEWEST SNAPSHOT: replication
INFO: Sending oldest full snapshot syncoid-test-0-0@replication (~ 347 KB) to new target filesystem:
 122KiB 0:00:00 [1,81MiB/s] [====================================================>                                                                                                    ] 35%            
+ zfs get encryptionroot -r -t filesystem syncoid-test-0-0 syncoid-test-0-1
NAME                          PROPERTY        VALUE                    SOURCE
syncoid-test-0-0              encryptionroot  syncoid-test-0-0         -
syncoid-test-0-0/a            encryptionroot  syncoid-test-0-0         -
syncoid-test-0-0/a/aa         encryptionroot  syncoid-test-0-0         -
syncoid-test-0-0/a/ab         encryptionroot  syncoid-test-0-0         -
syncoid-test-0-0/a/ac         encryptionroot  syncoid-test-0-0         -
syncoid-test-0-0/b            encryptionroot  syncoid-test-0-0         -
syncoid-test-0-0/c            encryptionroot  syncoid-test-0-0         -
syncoid-test-0-1              encryptionroot  -                        -
syncoid-test-0-1/backup       encryptionroot  syncoid-test-0-1/backup  -
syncoid-test-0-1/backup/a     encryptionroot  syncoid-test-0-1/backup  -
syncoid-test-0-1/backup/a/aa  encryptionroot  syncoid-test-0-1/backup  -
syncoid-test-0-1/backup/a/ab  encryptionroot  syncoid-test-0-1/backup  -
syncoid-test-0-1/backup/a/ac  encryptionroot  syncoid-test-0-1/backup  -
syncoid-test-0-1/backup/b     encryptionroot  syncoid-test-0-1/backup  -
syncoid-test-0-1/backup/c     encryptionroot  syncoid-test-0-1/backup  -
+ exit 0
+ cleanUp
+ zpool export syncoid-test-0-0
+ zpool export syncoid-test-0-1
```

**Comment by dani** [comment]
You're right, it's working. What I did for it to work 

- Use recursive = zfs in sanoid conf for the encrypted dataset
- use --no-sync-snap --sendopts="Rw' --recvopts="u" as argument to syncoid

the thing that misled me is that during the initial sync, the encryptionroot isn't correctly kept for sub datasets. Only after the initial sync is finished, zfs get encpryptionroot displays the correct one.

Thanks for your assistance :-)

**Comment by kyle0r** [comment]
@dani  @phreaker0 thanks for your knowledge share/capture here. Was very useful.

can one become a doc contributor? I could edit the docs to add this in.

**Comment by xrishox** [comment]
Do you only need to do sendoptions=Rw for the first send or for each time? I've found that there isn't a really convenient or quick way to send snapshots manually this way without manually creating the snapshot first. it keeps erroring out. After doing the initial replication over with sendoptions=Rw can I do it like this from there on out? "syncoid -R sendoptions=w" and still maintain encryptionroot?

**Comment by arpadmuller** [comment]
> * Use recursive = zfs in sanoid conf for the encrypted dataset
> * use --no-sync-snap --sendopts="Rw' --recvopts="u" as argument to syncoid

After several hours (days) of testing the same in plain zfs send/recv, you just saved me a bunch of time with this solution.

Thank you a lot!
Cheers!

**Comment by wardw** [comment]
Just a note that if you've already done the replication you can manually fixup the received dataset(s) to again inherit their key from their parent using the `-i` option to `zfs change-key`:

    zfs change-key -i zpool/backups/child

This will change the key of `zpool/backups/child` to inherit it's key from `zpool/backups`, and so restore the original pattern as on the host.

Being a manual fixup you'd have to do this again on receiving new datasets, but in fact just having the _option_ after the fact is sufficient for me to be no longer that concerned about this on the realization that it's not something you have to get right a-priori.  Just fixup later at any time.

And as an aside, it also means you can stick to `recursive = yes` in your `sanoid.conf` and let sanoid handle the recursion, which the docs also hint at as being preferred (although without much further elaboration).

**Comment by ossimoi** [comment]
> Just a note that if you've already done the replication you can manually fixup the received dataset(s) to again inherit their key from their parent using the `-i` option to `zfs change-key`:
> 
> ```
> zfs change-key -i zpool/backups/child
> ```
> 
> This will change the key of `zpool/backups/child` to inherit it's key from `zpool/backups`, and so restore the original pattern as on the host.
> 
> Being a manual fixup you'd have to do this again on receiving new datasets, but in fact just having the _option_ after the fact is sufficient for me to be no longer that concerned about this on the realization that it's not something you have to get right a-priori. Just fixup later at any time.
> 
> And as an aside, it also means you can stick to `recursive = yes` in your `sanoid.conf` and let sanoid handle the recursion, which the docs also hint at as being preferred (although without much further elaboration).

This indeed works, but the downside is that you have to unlock the root dataset at receiving end.

**Comment by wardw** [comment]
Yes you have to unlock the dataset and in fact I wanted to add this heads-up as it may be worse than that:

I've since learned that the use of `zfs change-key` on the destination dataset of a raw-send incremental stream is not such a good idea.  There seems to be some fairly long standing issues around the use of `change-key` and raw incremental streams.  It's not that datasets get corrupted exactly, but rather that datasets can fail to mount as a result of issues relating to the nature of the raw encrypted data and key metadata.  Depending on the changes it may be considerable work to undo the problem (and that zfs may not warn you, meaning it may be some time later before you realise you have a problem).

References below, but the short answer may be to simple avoid any use of `change-key` on the receive side of a dataset, or perhaps any use of `change-key` on a dataset that's participating in a raw-send incremental stream, preferring a new, full send after any key change instead.

+ https://github.com/openzfs/zfs/issues/12614
+ https://github.com/openzfs/zfs/issues/12123
+ https://sambowman.tech/blog/posts/mind-the-encryptionroot-how-to-save-your-data-when-zfs-loses-its-mind

In my case **I found I could no longer mount datasets on the receive side as a result of my previous comment**.  Fortunately these were backups and so I just started again with a new, full send.

**Comment by secabeen** [comment]
Honestly, I would broaden that and say that any `zfs change-key` action on a dataset with replicated copies on other systems should be approached with caution.  There are many actions that are fine to perform, both others can cause silent issues, as described.  If you are going to do a `change-key`, it would be best to test the entire workflow end-to-end on a test dataset first.
